File: python/fluent_cards/builders/column_set_builder.py
from __future__ import annotations
from typing import Callable, Optional, Union
from ..enums import ContainerStyle, HorizontalAlignment, Spacing


class ColumnSetBuilder:
    def __init__(self):
        self._column_set: dict = {'type': 'ColumnSet', 'columns': []}

    def with_id(self, id: str) -> ColumnSetBuilder:
        self._column_set['id'] = id
        return self

    def with_style(self, style: ContainerStyle) -> ColumnSetBuilder:
        self._column_set['style'] = style.value
        return self

    def with_bleed(self, bleed: bool = True) -> ColumnSetBuilder:
        self._column_set['bleed'] = bleed
        return self

    def with_min_height(self, min_height: str) -> ColumnSetBuilder:
        self._column_set['minHeight'] = min_height
        return self

    def with_horizontal_alignment(self, alignment: HorizontalAlignment) -> ColumnSetBuilder:
        self._column_set['horizontalAlignment'] = alignment.value
        return self

    def with_spacing(self, spacing: Spacing) -> ColumnSetBuilder:
        self._column_set['spacing'] = spacing.value
        return self

    def with_separator(self, separator: bool) -> ColumnSetBuilder:
        self._column_set['separator'] = separator
        return self

    def with_select_action(self, configure: Callable) -> ColumnSetBuilder:
        from .action_builder import ActionBuilder
        b = ActionBuilder()
        configure(b)
        self._column_set['selectAction'] = b.build()
        return self

    def add_column(self, configure_or_width: Union[str, Callable],
                   configure: Optional[Callable] = None) -> ColumnSetBuilder:
        from .column_builder import ColumnBuilder
        b = ColumnBuilder()
        if isinstance(configure_or_width, str):
            b.with_width(configure_or_width)
            configure(b)
        else:
            configure_or_width(b)
        self._column_set['columns'].append(b.build())
        return self

    def build(self) -> dict:
        return self._column_set

File: node/packages/fluent-cards/src/builders/ColumnSetBuilder.ts
import type { ColumnSet } from '../models.js';
import { ContainerStyle, HorizontalAlignment, Spacing } from '../enums.js';
import { ColumnBuilder } from './ColumnBuilder.js';
import { ActionBuilder } from './ActionBuilder.js';

/** Fluent builder for {@link ColumnSet} elements. */
export class ColumnSetBuilder {
  private readonly columnSet: ColumnSet = { type: 'ColumnSet', columns: [] };

  /** Sets the unique identifier. @param id The unique identifier. @returns The builder instance for method chaining. */
  withId(id: string): this { this.columnSet.id = id; return this; }
  /** Sets the container style. @param style The container style. @returns The builder instance for method chaining. */
  withStyle(style: ContainerStyle): this { this.columnSet.style = style; return this; }
  /** Sets whether the column set bleeds into the parent container. @param bleed True to enable bleed. @returns The builder instance for method chaining. */
  withBleed(bleed = true): this { this.columnSet.bleed = bleed; return this; }
  /** Sets the minimum height of the column set. @param minHeight The minimum height (e.g. `'100px'`). @returns The builder instance for method chaining. */
  withMinHeight(minHeight: string): this { this.columnSet.minHeight = minHeight; return this; }
  /** Sets the horizontal alignment of the columns. @param alignment The horizontal alignment. @returns The builder instance for method chaining. */
  withHorizontalAlignment(alignment: HorizontalAlignment): this { this.columnSet.horizontalAlignment = alignment; return this; }
  /** Sets the spacing above the element. @param spacing The spacing value. @returns The builder instance for method chaining. */
  withSpacing(spacing: Spacing): this { this.columnSet.spacing = spacing; return this; }
  /** Sets whether a separator line is displayed above the element. @param separator True to show a separator. @returns The builder instance for method chaining. */
  withSeparator(separator: boolean): this { this.columnSet.separator = separator; return this; }

  /** Sets the action to invoke when the column set is selected. @param configure A callback to configure the ActionBuilder. @returns The builder instance for method chaining. */
  withSelectAction(configure: (b: ActionBuilder) => void): this {
    const b = new ActionBuilder();
    configure(b);
    this.columnSet.selectAction = b.build();
    return this;
  }

  /** Adds a column using a configure callback. @param configure A callback to configure the Column builder. @returns The builder instance for method chaining. */
  addColumn(configure: (b: ColumnBuilder) => void): this;
  /** Adds a column with an explicit width. @param width The column width (e.g. `"auto"`, `"stretch"`, or a pixel value). @param configure A callback to configure the Column builder. @returns The builder instance for method chaining. */
  addColumn(width: string, configure: (b: ColumnBuilder) => void): this;
  addColumn(
    configureOrWidth: string | ((b: ColumnBuilder) => void),
    configure?: (b: ColumnBuilder) => void,
  ): this {
    let b = new ColumnBuilder();
    if (typeof configureOrWidth === 'string') {
      b.withWidth(configureOrWidth);
      configure!(b);
    } else {
      configureOrWidth(b);
    }
    this.columnSet.columns!.push(b.build());
    return this;
  }

  /** Builds and returns the configured ColumnSet. @returns The configured ColumnSet instance. */
  build(): ColumnSet { return this.columnSet; }
}
